Product Collection: Add 'Order By' Control to Product Collection Inspector

imanish003 opened the pull request:
Add 'Order By' Control to Product Collection Inspector.

<img width="289" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/011b91ec-868c-4c21-a91f-b8a02d025c62">


This PR adds a new 'Order By' control to the product collection inspector. The control allows users to specify the order of products by various attributes such as title and date. 

Fixes #9359 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

1. Create a new post
2. Add Product Collection block
3. Verify that the sidebar displays a 'Order By' setting when the Product Collection block is selected.
4. Change the order using the 'Order By' setting.
   - Confirm that the change is immediately reflected in the editor.
5. Save and publish the post.
   - Verify that the same change is reflected on the frontend.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add 'Order By' Control to Product Collection Inspector

imanish003 committed:
Add columns control to product collection block editor settings

- `InspectorControls` from './inspector-controls' is now imported in `edit.tsx` and used in the returned JSX of `Edit` function.
- A new file `columns-control.tsx` is added under 'product-collection' block's 'inspector-controls' directory which exports a `ColumnsControl` component. This component uses `RangeControl` from '@wordpress/components' to control the number of columns in the product collection display layout when the layout type is 'flex'.
- The types file (`types.ts`) for 'product-collection' block is updated. The `Attributes` interface is renamed to `ProductCollectionAttributes` and the `ProductCollectionContext` interface is removed. The `ProductCollectionAttributes` now includes 'queryContext', 'templateSlug', and 'displayLayout' properties.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

imanish003 committed:
Refactor: Simplify Fallback Return in ColumnsControl Component

This commit simplifies the fallback return value of the ColumnsControl component. Instead of returning an empty fragment (<> </>), it now returns null when the condition isn't met. This change improves readability and aligns with best practices for conditional rendering in React.

imanish003 committed:
Feature: Add 'Order By' Control to Product Collection Inspector

This commit adds a new 'Order By' control to the product collection inspector. The control allows users to specify the order of products in a collection by various attributes such as title and date. To support this, a new component 'OrderByControl' has been created and included in the product collection inspector. Additionally, the types for 'order' and 'orderBy' attributes have been updated and exported for reuse.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9359-product-collection-editor-settings-order-by

imanish003 commented:
Hi Karol,

In this pull request, I have added the following options in the Order By:
- A → Z
- Z → A
- Newest to Oldest
- Oldest to Newest
- Best Selling
- Top Rated

Currently, the last two options won't work because some changes are required in the `ProductTemplate.php` file. Since you are currently making a lot of changes in that file, I decided not to touch it for now to avoid conflicts. Please let me know once your pull request is merged, and I will make the necessary changes. 🤝

imanish003 committed:
Add more options to OrderBy type

imanish003 commented:
I have converted this pull request to a draft as there is one issue that I need to resolve. :)

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add orderby handling on frontend & editor

The main changes include:
1. Added a new property 'isProductCollectionBlock' in the block.json to denote if a block is a product collection block.
2. In the ProductCollection PHP class, a new initialization function has been defined to hook into the WordPress lifecycle, register the block, and update the query based on this block.
3. Added methods to manage query parameters for both frontend rendering and the Editor.
4. Expanded allowed 'collection_params' for the REST API to include custom 'orderby' values.
5. Defined a function to build the query based on block attributes, filters, and global WP_Query.
6. Created utility functions to handle complex query operations such as merging queries, handling custom sort values, and merging arrays recursively.

These improvements allow for more flexible and robust handling of product collections in both the front-end display and the WordPress editor. It also extends support for custom 'orderby' values in the REST API, which allows for more advanced sorting options in product collections.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 commented:
@kmanijak @nefeline Hey folks, just wanted to let you know that this is ready for review now. 🙂

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

kmanijak commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -22,4 +38,269 @@ class ProductCollection extends AbstractBlock {
+	 * @param array $base First array.
+	 * @param array $new  Second array.
+	 */
+	private function array_merge_recursive_replace_non_array_properties( $base, $new ) {
```
What do you think about extracting this method (and possibly others like `get_valid_query_vars`) that has exactly the same shape as in `ProductQuery` to the [`Utils`](https://github.com/woocommerce/woocommerce-blocks/tree/trunk/src/Utils), possibly [`Utils/BlocksWPQuery.php`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/Utils/BlocksWpQuery.php)?

If the method includes logic related to Product Collection (like:
```
		// If not in context of product collection block, return the query as is.
		if ( ! $block->context['query']['isProductCollectionBlock'] ) {
			return $query;
		}
```

then obviously not, but there are methods that could be shared.

imanish003 replied:
While I understand your perspective on refactoring this method and potentially others, such as get_valid_query_vars, into a Utils class like Utils/BlocksWPQuery.php, I would suggest caution for now. 

Should the need for these methods arise in other block contexts in the future, it would certainly make sense to generalize them in a Utils class. However, until that requirement is more evident, keeping these functions within the ProductCollection block may contribute to code readability and maintainability by preserving context.

Let's continue to monitor our needs as the codebase evolves. Refactoring for wider reuse can be undertaken when the value and necessity of doing so become clearer.

What do you think?

kmanijak replied:
I won't push for that, it was just a suggestion and I'm ok with your reasoning! 👍

imanish003 replied:
Thanks for your understanding and for being open to this perspective. Looking forward to more fruitful feedback in the future! 🚀

kmanijak reviewed (approved):
Great work! It works as expected! I left a minor comment, otherwise, it's all good, pre-approving.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
fix: handle undefined index for isProductCollectionBlock

This commit addresses a potential issue where the 'isProductCollectionBlock' index might not be defined in certain situations within the 'build_query' method of the ProductCollection class.

Previously, we directly accessed 'isProductCollectionBlock' from the 'query' context of the block. Now, we use the null coalescing operator (??) to ensure that we assign a default value of false if 'isProductCollectionBlock' is not set.

This change provides a safer way to handle the scenario when the 'isProductCollectionBlock' is not defined in the block context and helps prevent undefined index warnings.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by
